File: sdk/revdist/typescript/revdist/oracle.ts
/** SOL/2Z oracle client. */

export interface SwapRate {
  rate: number;
  timestamp: number;
  signature: string;
  solPriceUsd: string;
  twozPriceUsd: string;
  cacheHit: boolean;
}

export class OracleClient {
  private readonly baseUrl: string;

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  async fetchSwapRate(): Promise<SwapRate> {
    const resp = await fetch(`${this.baseUrl}/swap-rate`);
    if (!resp.ok) {
      throw new Error(`oracle returned status ${resp.status}`);
    }
    const data = await resp.json();
    return {
      rate: data.swapRate,
      timestamp: data.timestamp,
      signature: data.signature,
      solPriceUsd: data.solPriceUsd,
      twozPriceUsd: data.twozPriceUsd,
      cacheHit: data.cacheHit,
    };
  }
}

File: sdk/revdist/python/revdist/oracle.py
"""SOL/2Z oracle client."""

from __future__ import annotations

from dataclasses import dataclass

import httpx


@dataclass
class SwapRate:
    rate: float
    timestamp: int
    signature: str
    sol_price_usd: str
    twoz_price_usd: str
    cache_hit: bool


class OracleClient:
    """Fetches SOL/2Z swap rates from the oracle API."""

    def __init__(self, base_url: str) -> None:
        self._base_url = base_url
        self._http = httpx.Client(timeout=30)

    def fetch_swap_rate(self) -> SwapRate:
        resp = self._http.get(f"{self._base_url}/swap-rate")
        resp.raise_for_status()
        data = resp.json()
        return SwapRate(
            rate=data["swapRate"],
            timestamp=data["timestamp"],
            signature=data["signature"],
            sol_price_usd=data["solPriceUsd"],
            twoz_price_usd=data["twozPriceUsd"],
            cache_hit=data["cacheHit"],
        )
